File: src/org/genyris/dl/GraphFunction.java
//Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.dl;

import org.genyris.core.Bignum;
import org.genyris.core.Constants;
import org.genyris.core.Exp;
import org.genyris.core.Symbol;
import org.genyris.exception.GenyrisException;
import org.genyris.interp.AbstractMethod;
import org.genyris.interp.ApplicableFunction;
import org.genyris.interp.Closure;
import org.genyris.interp.Environment;
import org.genyris.interp.Interpreter;
import org.genyris.interp.UnboundException;

public class GraphFunction extends ApplicableFunction {

	public GraphFunction(Interpreter interp) {
		super(interp, "graph", true);
	}

	public Exp bindAndExecute(Closure proc, Exp[] arguments,
			Environment envForBindOperations) throws GenyrisException {
		AbstractGraph ts = new GraphHashSimple();
		for (int i = 0; i < arguments.length; i++) {
			addTripleFromList(ts, arguments[i]);
		}
		return ts;
	}

	private void addTripleFromList(AbstractGraph ts, Exp exp)
			throws GenyrisException {
		Symbol subject = toSymbol(exp.car());
        Symbol predicate = toSymbol(exp.cdr().car());
        Exp object = exp.cdr().cdr().car();

		ts.add(new Triple(subject, predicate, object));
	}

	public static void bindFunctionsAndMethods(Interpreter interpreter)
			throws UnboundException, GenyrisException {
		interpreter.bindGlobalProcedureInstance(new GraphFunction(
				interpreter));
		interpreter.bindMethodInstance(Constants.GRAPH, new AddMethod(
				interpreter));
		interpreter.bindMethodInstance(Constants.GRAPH, new SelectMethod(
				interpreter));
		interpreter.bindMethodInstance(Constants.GRAPH,
				new AsTriplesMethod(interpreter));
		interpreter.bindMethodInstance(Constants.GRAPH, new RemoveMethod(
				interpreter));
		interpreter.bindMethodInstance(Constants.GRAPH, new LengthMethod(
				interpreter));
		interpreter.bindMethodInstance(Constants.GRAPH, new GetMethod(
				interpreter));
		interpreter.bindMethodInstance(Constants.GRAPH, new PutMethod(
				interpreter));
		interpreter.bindMethodInstance(Constants.GRAPH,
				new GetListMethod(interpreter));
		interpreter.bindMethodInstance(Constants.GRAPH,
				new SubjectsMethod(interpreter));
        interpreter.bindMethodInstance(Constants.GRAPH,
                new UnionMethod(interpreter));
	}

	public abstract static class AbstractGraphMethod extends
			AbstractMethod {

		public AbstractGraphMethod(Interpreter interp, String name) {
			super(interp, name);
		}

		protected AbstractGraph getSelfGraph(Environment env)
				throws GenyrisException {
			getSelf(env);
			if (!(_self instanceof AbstractGraph)) {
				throw new GenyrisException(
						"Non-Graph passed to a Graph method.");
			} else {
				return (AbstractGraph) _self;
			}
		}
	}

	public static class AddMethod extends AbstractGraphMethod {

		public AddMethod(Interpreter interp) {
			super(interp, "add");
		}

		public Exp bindAndExecute(Closure proc, Exp[] arguments, Environment env)
				throws GenyrisException {
			AbstractGraph self = getSelfGraph(env);
			checkArguments(arguments, 1);
			Class[] types = { Triple.class };
			checkArgumentTypes(types, arguments);
			self.add((Triple) arguments[0]);
			return _self;
		}
	}

	public static class GetMethod extends AbstractGraphMethod {

		public GetMethod(Interpreter interp) {
			super(interp, "get");
		}

		public Exp bindAndExecute(Closure proc, Exp[] arguments, Environment env) throws GenyrisException {
			AbstractGraph self = getSelfGraph(env);
			checkArguments(arguments, 2);
			Class[] types = { Exp.class, Exp.class };
			checkArgumentTypes(types, arguments);
			return self.get(toSymbol(arguments[0]), toSymbol(arguments[1]));
		}
	}

	public static class GetListMethod extends AbstractGraphMethod {

		public GetListMethod(Interpreter interp) {
            super(interp, "get-list");
        }

		public Exp bindAndExecute(Closure proc, Exp[] arguments, Environment env)
				throws GenyrisException {
			AbstractGraph self = getSelfGraph(env);
			checkArguments(arguments, 2);
			Class[] types = { Exp.class, Exp.class };
			checkArgumentTypes(types, arguments);
			return self.getList(toSymbol(arguments[0]), toSymbol(arguments[1]), NIL);
		}
	}

	public static class PutMethod extends AbstractGraphMethod {

		public PutMethod(Interpreter interp) {
			super(interp, "put");
		}

		public Exp bindAndExecute(Closure proc, Exp[] arguments, Environment env)
				throws GenyrisException {
			AbstractGraph self = getSelfGraph(env);
			checkArguments(arguments, 3);
			Class[] types = { Exp.class, Exp.class, Exp.class };
			checkArgumentTypes(types, arguments);
			self.put(toSymbol(arguments[0]), toSymbol(arguments[1]), arguments[2]);
			return _self;
		}
	}

    public static class SubjectsMethod extends AbstractGraphMethod {

        public SubjectsMethod(Interpreter interp) {
            super(interp, "subjects");
        }

        public Exp bindAndExecute(Closure proc, Exp[] arguments, Environment env)
                throws GenyrisException {
            AbstractGraph self = getSelfGraph(env);
            return self.subjects(NIL);
        }
    }

	public static class UnionMethod extends AbstractGraphMethod {

		public UnionMethod(Interpreter interp) {
			super(interp, "union");
		}

		public Exp bindAndExecute(Closure proc, Exp[] arguments, Environment env)
				throws GenyrisException {
            checkArguments(arguments, 1);
            Class[] types = { AbstractGraph.class };
            checkArgumentTypes(types, arguments);
			AbstractGraph self = getSelfGraph(env);
			AbstractGraph other = (AbstractGraph)arguments[0]; 
			return self.union(other);
		}
	}

	public static class LengthMethod extends AbstractGraphMethod {

		public LengthMethod(Interpreter interp) {
			super(interp, "length");
		}

		public Exp bindAndExecute(Closure proc, Exp[] arguments, Environment env)
				throws GenyrisException {
			AbstractGraph self = getSelfGraph(env);
			return new Bignum(self.length());
		}
	}

	public static class SelectMethod extends AbstractGraphMethod {

		public SelectMethod(Interpreter interp) {
			super(interp, "select");
		}

		public Exp bindAndExecute(Closure proc, Exp[] arguments, Environment env)
				throws GenyrisException {
			Closure closure = null;
			AbstractGraph self = getSelfGraph(env);
			checkMinArguments(arguments, 3);
			Class[] types = { Exp.class, Exp.class, Exp.class };
			checkArgumentTypes(types, arguments);
			Symbol subject = toSymbol(arguments[0]);
			Symbol predicate = toSymbol(arguments[1]);
			Exp object = arguments[2];
			if (arguments[0] == NIL) {
				subject = null;
			}
			if (arguments[1] == NIL) {
				predicate = null;
			}
			if (arguments[2] == NIL) {
				object = null;
			}
/* #TODO			if (arguments.length == 4 && arguments[3] != NIL) {
				closure = (Closure) arguments[3];
			}
*/
			return self.select(subject, predicate, object, closure, env);
		}
	}

	public static class AsTriplesMethod extends AbstractGraphMethod {

		public AsTriplesMethod(Interpreter interp) {
			super(interp, "asTriples");
		}

		public Exp bindAndExecute(Closure proc, Exp[] arguments, Environment env)
				throws GenyrisException {
			AbstractGraph self = getSelfGraph(env);
			return self.asTripleList(NIL);
		}
	}

	public static class RemoveMethod extends AbstractGraphMethod {

		public RemoveMethod(Interpreter interp) {
			super(interp, "remove");
		}

		public Exp bindAndExecute(Closure proc, Exp[] arguments, Environment env)
				throws GenyrisException {
			AbstractGraph self = getSelfGraph(env);
			checkArguments(arguments, 1);
			Class[] types = { Triple.class };
			checkArgumentTypes(types, arguments);
			self.remove((Triple) arguments[0]);
			return self;
		}
	}

}

File: src/org/genyris/dl/GraphFunction.kt
//Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.dl

import org.genyris.core.*
import org.genyris.exception.GenyrisException
import org.genyris.interp.*

class GraphFunction(interp: Interpreter) : ApplicableFunction(interp, "graph", true) {
    private val _tripleSymbol: Symbol?

    init {
        _tripleSymbol = _interp.intern("triple")
    }

    @kotlin.Throws(GenyrisException::class)
    override fun bindAndExecute(
        proc: Closure?, arguments: Array<Exp?>,
        envForBindOperations: Environment?
    ): Exp {
        val ts: AbstractGraph = GraphHashSimple()
        for (i in arguments.indices) {
            addTripleFromList(ts, arguments[i]!!)
        }
        return ts
    }

    @kotlin.Throws(GenyrisException::class)
    private fun addTripleFromList(ts: AbstractGraph, exp: Exp) {
        // Accept either (s p o) or ^(triple s p o) or a Triple
        // also accept (s = o) -> (triple nil s 0)
        when (exp) {
            -> {
                if (ep.length(NIL) == 1 && ep.cdr() !== NIL) { // (s = o)
                    if (ep.car() !is Symbol) {
                        throw GenyrisException("triple predicate must be a symbol in: " + exp.toString())
                    }
                    ts.add(Triple(NIL, ep.car() as Symbol?, ep.cdr()))
                    return
                }
                if (ep.length(NIL) == 4 && ep.car() == _tripleSymbol) {
                    ep = ep.cdr() as Pair?
                }
                val subject = toSymbol(ep.car())
                val predicate = toSymbol(ep.cdr().car())
                val `object`: Exp? = ep.cdr().cdr().car()
                ts.add(Triple(subject, predicate, `object`))
            }

            -> ts.add(t)
            else -> throw GenyrisException("unknown type in graph argument:" + exp.toString())
        }
    }

    abstract class AbstractGraphMethod(interp: Interpreter?, name: String?) : AbstractMethod(interp, name) {
        @kotlin.Throws(GenyrisException::class)
        protected fun getSelfGraph(env: Environment): AbstractGraph {
            getSelf(env)
            if (_self !is AbstractGraph) {
                throw GenyrisException(
                    "Non-Graph passed to a Graph method."
                )
            } else {
                return _self as AbstractGraph
            }
        }
    }

    class AddMethod(interp: Interpreter?) : AbstractGraphMethod(interp, "add") {
        @kotlin.Throws(GenyrisException::class)
        override fun bindAndExecute(proc: Closure?, arguments: Array<Exp>, env: Environment): Exp? {
            val self = getSelfGraph(env)
            checkArguments(arguments, 1, 3)
            if (arguments.size == 1) {
                val types = arrayOf<Class<*>?>(Triple::class.java)
                checkArgumentTypes(types, arguments)
                self.add(arguments[0] as Triple?)
                return _self
            }

            checkArguments(arguments, 2, 3)
            val types = arrayOf<Array<Class<*>?>?>(
                arrayOf<Class<*>?>(Symbol::class.java),
                arrayOf<Class<*>?>(Symbol::class.java, Pair::class.java),
                arrayOf<Class<*>?>(
                    Exp::class.java
                )
            )
            checkSuppliedArgumentTypes(types, arguments)
            if (arguments[1] is Pair) {
                var head = arguments[1]
                while (head !== NIL) {
                    if (head.car() !is Pair) {
                        throw GenyrisException("non-pair in .add predicate " + head.car())
                    }
                    self.add(Triple(toSymbol(arguments[0]), toSymbol(head.car().car()), head.car().cdr()))
                    head = head.cdr()
                }
                return _self
            }
            self.add(Triple(toSymbol(arguments[0]), toSymbol(arguments[1]), arguments[2]))
            return _self
        }
    }

    class GetMethod(interp: Interpreter?) : AbstractGraphMethod(interp, "get") {
        @kotlin.Throws(GenyrisException::class)
        override fun bindAndExecute(proc: Closure?, arguments: Array<Exp>, env: Environment): Exp? {
            val self = getSelfGraph(env)
            checkArguments(arguments, 2)
            val types = arrayOf<Class<*>?>(Exp::class.java, Exp::class.java)
            checkArgumentTypes(types, arguments)
            return self.get(toSymbol(arguments[0]), toSymbol(arguments[1]))
        }
    }

    class GetListMethod(interp: Interpreter?) : AbstractGraphMethod(interp, "get-list") {
        @kotlin.Throws(GenyrisException::class)
        override fun bindAndExecute(proc: Closure?, arguments: Array<Exp>, env: Environment): Exp? {
            val self = getSelfGraph(env)
            checkArguments(arguments, 2)
            val types = arrayOf<Class<*>?>(Exp::class.java, Exp::class.java)
            checkArgumentTypes(types, arguments)
            return self.getList(toSymbol(arguments[0]), toSymbol(arguments[1]), NIL)
        }
    }

    class PutMethod(interp: Interpreter?) : AbstractGraphMethod(interp, "put") {
        @kotlin.Throws(GenyrisException::class)
        override fun bindAndExecute(proc: Closure?, arguments: Array<Exp>, env: Environment): Exp? {
            val self = getSelfGraph(env)
            checkArguments(arguments, 3)
            val types = arrayOf<Class<*>?>(Symbol::class.java, Symbol::class.java, Exp::class.java)
            checkArgumentTypes(types, arguments)
            self.put(toSymbol(arguments[0]), toSymbol(arguments[1]), arguments[2])
            return _self
        }
    }

    class SubjectsMethod(interp: Interpreter?) : AbstractGraphMethod(interp, "subjects") {
        @kotlin.Throws(GenyrisException::class)
        override fun bindAndExecute(proc: Closure?, arguments: Array<Exp?>?, env: Environment): Exp? {
            val self = getSelfGraph(env)
            return self.subjects(NIL)
        }
    }

    class PredicatesMethod(interp: Interpreter?) : AbstractGraphMethod(interp, "predicates") {
        @kotlin.Throws(GenyrisException::class)
        override fun bindAndExecute(proc: Closure?, arguments: Array<Exp>, env: Environment): Exp? {
            checkArguments(arguments, 1)
            val types = arrayOf<Class<*>?>(Symbol::class.java)
            checkArgumentTypes(types, arguments)
            val self = getSelfGraph(env)
            val subject = arguments[0] as Symbol?
            return self.predicates(subject, NIL)
        }
    }

    class UnionMethod(interp: Interpreter?) : AbstractGraphMethod(interp, "union") {
        @kotlin.Throws(GenyrisException::class)
        override fun bindAndExecute(proc: Closure?, arguments: Array<Exp>, env: Environment): Exp? {
            checkArguments(arguments, 1)
            val types = arrayOf<Class<*>?>(AbstractGraph::class.java)
            checkArgumentTypes(types, arguments)
            val self = getSelfGraph(env)
            val other = arguments[0] as AbstractGraph?
            return self.union(other)
        }
    }

    class LengthMethod(interp: Interpreter?) : AbstractGraphMethod(interp, "length") {
        @kotlin.Throws(GenyrisException::class)
        override fun bindAndExecute(proc: Closure?, arguments: Array<Exp?>?, env: Environment): Exp {
            val self = getSelfGraph(env)
            return Bignum(self.length())
        }
    }

    class SelectMethod(interp: Interpreter?) : AbstractGraphMethod(interp, "select") {
        @kotlin.Throws(GenyrisException::class)
        override fun bindAndExecute(proc: Closure?, arguments: Array<Exp>, env: Environment): Exp? {
            val closure: Closure? = null
            val self = getSelfGraph(env)
            checkMinArguments(arguments, 3)
            val types = arrayOf<Class<*>?>(Exp::class.java, Exp::class.java, Exp::class.java)
            checkArgumentTypes(types, arguments)
            var subject = toSymbol(arguments[0])
            var predicate = toSymbol(arguments[1])
            var `object`: Exp? = arguments[2]
            if (arguments[0] === NIL) {
                subject = null
            }
            if (arguments[1] === NIL) {
                predicate = null
            }
            if (arguments[2] === NIL) {
                `object` = null
            }
            /* #TODO			if (arguments.length == 4 && arguments[3] != NIL) {
				closure = (Closure) arguments[3];
			}
*/
            return self.select(subject, predicate, `object`, closure, env)
        }
    }

    class AsTriplesMethod(interp: Interpreter?) : AbstractGraphMethod(interp, "asTriples") {
        @kotlin.Throws(GenyrisException::class)
        override fun bindAndExecute(proc: Closure?, arguments: Array<Exp?>?, env: Environment): Exp? {
            val self = getSelfGraph(env)
            return self.asTripleList(NIL)
        }
    }

    class RemoveMethod(interp: Interpreter?) : AbstractGraphMethod(interp, "remove") {
        @kotlin.Throws(GenyrisException::class)
        override fun bindAndExecute(proc: Closure?, arguments: Array<Exp>, env: Environment): Exp {
            val self = getSelfGraph(env)
            checkArguments(arguments, 1)
            val types = arrayOf<Class<*>?>(Triple::class.java)
            checkArgumentTypes(types, arguments)
            self.remove(arguments[0] as Triple?)
            return self
        }
    }

    companion object {
        @kotlin.Throws(UnboundException::class, GenyrisException::class)
        fun bindFunctionsAndMethods(interpreter: Interpreter) {
            interpreter.bindGlobalProcedureInstance(GraphFunction(interpreter))
            interpreter.bindMethodInstance(Constants.GRAPH, AddMethod(interpreter))
            interpreter.bindMethodInstance(Constants.GRAPH, AsTriplesMethod(interpreter))
            interpreter.bindMethodInstance(Constants.GRAPH, GetListMethod(interpreter))
            interpreter.bindMethodInstance(Constants.GRAPH, GetMethod(interpreter))
            interpreter.bindMethodInstance(Constants.GRAPH, LengthMethod(interpreter))
            interpreter.bindMethodInstance(Constants.GRAPH, PredicatesMethod(interpreter))
            interpreter.bindMethodInstance(Constants.GRAPH, PutMethod(interpreter))
            interpreter.bindMethodInstance(Constants.GRAPH, RemoveMethod(interpreter))
            interpreter.bindMethodInstance(Constants.GRAPH, SelectMethod(interpreter))
            interpreter.bindMethodInstance(Constants.GRAPH, SubjectsMethod(interpreter))
            interpreter.bindMethodInstance(Constants.GRAPH, UnionMethod(interpreter))
        }
    }
}
